[CIR][X86] Implement lowering for `readwritebarrier` builtins

Luhaocong opened the pull request:
- Support CIR codegen for follow X86 builtin: `_ReadWriteBarrier` `_ReadBarrier` `_WriteBarrier` `__faststorefence`
- CIR dialect operations may have no results, no values will be returned after `emitTargetBuiltinExpr` even if it executes successfully. Using `std::optional<mlir::Value>` as return type to reslove this problem.
- Part of [#167765](https://github.com/llvm/llvm-project/issues/167765)

Lancern reviewed (approved):
LGTM in general, some nit comments inline.

andykaylor reviewed (commented):
I think this is OK, but we'd need some cleanup around the optional return type when we've already issued diagnostics.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -1523,13 +1523,37 @@ mlir::Value CIRGenFunction::emitX86BuiltinExpr(unsigned builtinID,
+        cir::MemOrder::SequentiallyConsistent,
+        cir::SyncScopeKindAttr::get(&getMLIRContext(),
+                                    cir::SyncScopeKind::System));
+    return mlir::Value{};
```
Indeed! Missed that.

Luhaocong commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -1352,7 +1352,14 @@ RValue CIRGenFunction::emitBuiltinExpr(const GlobalDecl &gd, unsigned builtinID,
   // Now see if we can emit a target-specific builtin.
-  if (mlir::Value v = emitTargetBuiltinExpr(builtinID, e, returnValue)) {
+  if (std::optional<mlir::Value> rst =
+          emitTargetBuiltinExpr(builtinID, e, returnValue)) {
```
Thanks, I put your suggestions in the annotation and apply them to the relevant code changes.

andykaylor reviewed (commented):
You should hold this until https://github.com/llvm/llvm-project/pull/171248 is commited and then rebase this with that change. Otherwise, this looks good.

Luhaocong committed:
[CIR][X86] Implement lowering for `readwritebarrier` builtins

- Support CIR codegen for follow X86 builtin:
  `_ReadWriteBarrier`
  `_ReadBarrier`
  `_WriteBarrier`
  `__faststorefence`
- CIR dialect operations maybe have no results, no value will return
  after `emitTargetBuiltinExpr` even if it executes successfully.
  Using `std::optional<mlir::Value>` as return type to reslove this
  problem.

andykaylor reviewed (approved):
lgtm
